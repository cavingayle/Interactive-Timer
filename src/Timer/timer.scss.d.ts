// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'button-container': string;
  'center': string;
  'container': string;
  'reset-button': string;
  'selected-speed-button': string;
  'speed-button': string;
  'start-button': string;
  'tagline': string;
  'time-display': string;
  'timer': string;
  'timer-input': string;
  'timer-input-container': string;
}
export const cssExports: CssExports;
export default cssExports;
