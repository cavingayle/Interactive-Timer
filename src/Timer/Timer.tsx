import React, { ChangeEvent, useState, useEffect, useCallback } from 'react'
import styles from './timer.scss'

import { AiOutlinePauseCircle } from 'react-icons/ai'

enum TimerSpeed {
  oneSecond = 1000,
  twoThirdsSecond = 1000 / 1.5,
  halfSecond = 1000 / 2,
}

const Timer = () => {
  const [currentInterval, setCurrentInterval] = useState<NodeJS.Timer>()
  const [isRunning, setIsRunning] = useState<boolean>()

  const [time, setTime] = useState<number>()
  const [timeInput, setTimeInput] = useState<number>()
  const [timerSpeed, setTimerSpeed] = useState<TimerSpeed>(TimerSpeed.oneSecond)

  const formatTime = (seconds: number) => {
    return (
      (seconds - (seconds %= 60)) / 60 + (seconds > 9 ? ':' : ':0') + seconds
    )
  }

  const handleInputChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const input = +e.target.value.replace(/\D/g, '')
      setTimeInput(input)
      if (!isRunning) {
        setTime(input * 60)
      }
    },
    [timeInput, time]
  )

  const startTimer = useCallback(
    (timerSpeed: any) => {
      if (!time || time === 0) {
        return
      }
      setIsRunning(true)
      const interval = setInterval(() => {
        setTime((prevInput) => {
          if (prevInput === 1) {
            clearInterval(interval)
            setIsRunning(false)
          }
          return prevInput - 1
        })
      }, timerSpeed)
      setCurrentInterval(interval)
    },
    [timeInput, isRunning, time, timerSpeed]
  )

  const pauseTimer = () => {
    clearInterval(currentInterval)
    setIsRunning(false)
  }

  const formattedTime = formatTime(time)

  return (
    <div className={styles.container}>
      <div className={styles.timer}>
        <span className={styles['timer-input-container']}>
          <strong>Countdown:</strong>
          <input
            className={styles['timer-input']}
            placeholder="(Min)"
            onChange={handleInputChange}
            min={0}
            type="tel"
            max={999}
            value={timeInput}
            maxLength={3}
            aria-label="timer-input"
          />
          <button className={styles['start-button']} onClick={startTimer} aria-label="timer-start">
            Start
          </button>
        </span>
        <span className={styles.tagline}>More than halfway there!</span>
        <div className={styles.center}>
          {' '}
          <span className={styles['time-display']} aria-label="timer-display">
            {' '}
            {time ? formattedTime : '00:00'}
          </span>
          <span onClick={pauseTimer}>
            <AiOutlinePauseCircle size={40} />
          </span>
        </div>

        <span className={styles['timer-input-container']}>
          <button
            className={
              timerSpeed === TimerSpeed.oneSecond
                ? styles['selected-speed-button']
                : styles['speed-button']
            }
            onClick={() => setTimerSpeed(TimerSpeed.oneSecond)}
          >
            1x
          </button>
          <button
            className={
              timerSpeed === TimerSpeed.twoThirdsSecond
                ? styles['selected-speed-button']
                : styles['speed-button']
            }
            onClick={() => setTimerSpeed(TimerSpeed.twoThirdsSecond)}
          >
            1.5
          </button>
          <button
            className={
              timerSpeed === TimerSpeed.halfSecond
                ? styles['selected-speed-button']
                : styles['speed-button']
            }
            onClick={() => setTimerSpeed(TimerSpeed.halfSecond)}
          >
            2x
          </button>
        </span>
      </div>
    </div>
  )
}

export default Timer
