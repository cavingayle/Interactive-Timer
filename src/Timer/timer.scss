@import url('https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,300;0,400;0,500;0,600;0,700;0,800;1,300;1,400;1,500;1,600;1,700;1,800&display=swap');

body {
  font-family: 'Open Sans', sans-serif;
}
.container {
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
}

.timer {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.timer-input {
  width: 100px;
  border-radius: none;
  padding: 12px 20px;
  margin: 8px 0;
  box-sizing: border-box;
}

.timer-input-container {
  // display: flex;
  // justify-content: space-between;
  // width: 320px;
  > * {
    margin: 0 5px;
  }
}

.timer-input-container {
}

.button-container {
  display: flex;
  align-items: center;
  justify-content: center;
}
.tagline {
  margin: 20px 0 0 0;
}
.time-display {
  font-size: 148px;
}

.center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.start-button {
  padding: 12px 20px;
  background-color: #0abba0;
  color: white;
  width: 100px;
  border: 1px solid #0abba0;
}

.reset-button {
  padding: 12px 20px;
  background-color: #53aff4;
  color: white;
  width: 100px;
  border: 1px solid #53aff4;
}

.disabled-button {
  padding: 12px 20px;
  background-color: gray;
  color: white;
  width: 100px;
  border: 1px solid gray;
}

.speed-button {
  border: 1px solid black;
  padding: 12px 20px;
}

.selected-speed-button {
  background: gray;
  color: white;
  padding: 12px 20px;
  border: 1px solid black;
}
