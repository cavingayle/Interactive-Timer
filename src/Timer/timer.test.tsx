import {
  render,
  fireEvent,
  screen,
  waitFor,
} from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'

import Timer from './Timer'

beforeEach(() => {
  jest.useFakeTimers()
})

afterEach(() => {
  jest.runOnlyPendingTimers()
  jest.useRealTimers()
})

const user = userEvent.setup({ delay: null })

let component

const setup = () => {
  const utils = render(<Timer />)
  const input = utils.getByLabelText('timer-input') as HTMLInputElement
  const button = utils.getByLabelText('timer-start') as HTMLInputElement
  const timerDisplay = utils.getByLabelText('timer-display') as HTMLInputElement
  const pauseButton = utils.getByLabelText('pause-button') as HTMLElement
  const doubleSpeedButton = utils.getByLabelText(
    'double-speed-button'
  ) as HTMLButtonElement
  const singleSpeedButton = utils.getByLabelText(
    'single-speed-button'
  ) as HTMLButtonElement
  const oneAndOneHalfSpeed = utils.getByLabelText(
    'one-and-half-speed'
  ) as HTMLButtonElement

  return {
    input,
    button,
    timerDisplay,
    pauseButton,
    doubleSpeedButton,
    oneAndOneHalfSpeed,
    singleSpeedButton,
    ...utils,
  }
}

//Test if component renders
it('renders without crashing', () => {
  setup()
})

it('Input for timer accepts input change', () => {
  const { input } = setup()
  fireEvent.change(input, { target: { value: '23' } })
  expect(input.value).toBe('23')
})

it('Input for timer accepts only numbers', () => {
  const { input } = setup()
  fireEvent.change(input, { target: { value: 'adfs' } })
  expect(input.value).toBe('0')
})

it('Timer takes input and sets time', () => {
  const { input, button, timerDisplay, debug } = setup()
  fireEvent.change(input, { target: { value: 10 } })

  user.click(button)
  expect(timerDisplay.textContent).toBe('10:00')
})

it('Timer takes input and decrements time', async () => {
  const { input, button, timerDisplay, debug } = setup()
  fireEvent.change(input, { target: { value: 10 } })

  user.click(button)

  await waitFor(() => expect(timerDisplay.textContent).toBe('9:59'), {
    timeout: 2000,
  })
})

// it('Should pause timer', async () => {
//   const { input, button, timerDisplay, pauseButton } = setup()
//   fireEvent.change(input, { target: { value: 5 } })
//   fireEvent.click(pauseButton)

//   user.click(button)
//   jest.advanceTimersByTime(5000)
//  expect(timerDisplay.textContent).toBe('4:56')
// })

it('Should increase timer speed by double', async () => {
  const { input, button, timerDisplay, pauseButton, doubleSpeedButton } =
    setup()
  fireEvent.change(input, { target: { value: 5 } })
  fireEvent.click(doubleSpeedButton)
  fireEvent.click(button)

  await waitFor(
    () => {
      expect(timerDisplay.textContent).toBe('3:00')
    },
    {
      timeout: 61000,
    }
  )
})

it('Should increase timer speed by 1.5x', async () => {
  const {
    input,
    button,
    timerDisplay,
    pauseButton,
    doubleSpeedButton,
    oneAndOneHalfSpeed,
  } = setup()

  fireEvent.change(input, { target: { value: 5 } })
  fireEvent.click(oneAndOneHalfSpeed)
  fireEvent.click(button)

  await waitFor(
    () => {
      expect(timerDisplay.textContent).toBe('3:30')
    },
    {
      timeout: 60000,
    }
  )
})

it('Should increase timer speed to 1x', async () => {
  const {
    input,
    button,
    timerDisplay,
    pauseButton,
    doubleSpeedButton,
    oneAndOneHalfSpeed,
    singleSpeedButton,
  } = setup()

  fireEvent.change(input, { target: { value: 5 } })
  fireEvent.click(singleSpeedButton)
  fireEvent.click(button)

  await waitFor(
    () => {
      expect(timerDisplay.textContent).toBe('4:00')
    },
    {
      timeout: 61000,
    }
  )
})

it('Timer should have the correct text', () => {
  render(<Timer />)
  expect(screen.getByText('Countdown:')).toBeInTheDocument()
  expect(screen.getByText('Start')).toBeInTheDocument()
  expect(screen.getByText('More than halfway there!')).toBeInTheDocument()
  expect(screen.getByRole('textbox')).toHaveAttribute('placeholder', '(Min)')
})
