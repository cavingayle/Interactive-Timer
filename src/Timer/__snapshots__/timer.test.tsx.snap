// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render the page and match the snapshot 1`] = `
<div>
  <div>
    <span>
      <strong>
        Countdown:
      </strong>
      <input
        aria-label="timer-input"
        max={999}
        maxLength={3}
        min={0}
        onChange={[Function]}
        placeholder="(Min)"
        type="tel"
      />
      <button
        aria-label="timer-start"
        onClick={[Function]}
      >
        Start
      </button>
    </span>
    <span>
      More than halfway there!
    </span>
    <div>
       
      <span
        aria-label="timer-display"
      >
        00:00
      </span>
      <span
        aria-label="pause-button"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={40}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </div>
    <span>
      <button
        aria-label="single-speed-button"
        onClick={[Function]}
      >
        1x
      </button>
      <button
        aria-label="one-and-half-speed"
        onClick={[Function]}
      >
        1.5
      </button>
      <button
        aria-label="double-speed-button"
        onClick={[Function]}
      >
        2x
      </button>
    </span>
  </div>
</div>
`;

exports[`Should render the page and match the snapshot 1`] = `
<div>
  <div>
    <span>
      <strong>
        Countdown:
      </strong>
      <input
        aria-label="timer-input"
        max={999}
        maxLength={3}
        min={0}
        onChange={[Function]}
        placeholder="(Min)"
        type="tel"
      />
      <button
        aria-label="timer-start"
        onClick={[Function]}
      >
        Start
      </button>
    </span>
    <span>
      More than halfway there!
    </span>
    <div>
       
      <span
        aria-label="timer-display"
      >
        00:00
      </span>
      <span
        aria-label="pause-button"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={40}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </div>
    <span>
      <button
        aria-label="single-speed-button"
        onClick={[Function]}
      >
        1x
      </button>
      <button
        aria-label="one-and-half-speed"
        onClick={[Function]}
      >
        1.5
      </button>
      <button
        aria-label="double-speed-button"
        onClick={[Function]}
      >
        2x
      </button>
    </span>
  </div>
</div>
`;

exports[`Should render with all page elements 1`] = `
<div>
  <div>
    <span>
      <strong>
        Countdown:
      </strong>
      <input
        aria-label="timer-input"
        max={999}
        maxLength={3}
        min={0}
        onChange={[Function]}
        placeholder="(Min)"
        type="tel"
      />
      <button
        aria-label="timer-start"
        onClick={[Function]}
      >
        Start
      </button>
    </span>
    <span>
      More than halfway there!
    </span>
    <div>
       
      <span
        aria-label="timer-display"
      >
        00:00
      </span>
      <span
        aria-label="pause-button"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={40}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </div>
    <span>
      <button
        aria-label="single-speed-button"
        onClick={[Function]}
      >
        1x
      </button>
      <button
        aria-label="one-and-half-speed"
        onClick={[Function]}
      >
        1.5
      </button>
      <button
        aria-label="double-speed-button"
        onClick={[Function]}
      >
        2x
      </button>
    </span>
  </div>
</div>
`;

exports[`renders 1`] = `
<div>
  <div>
    <span>
      <strong>
        Countdown:
      </strong>
      <input
        aria-label="timer-input"
        max={999}
        maxLength={3}
        min={0}
        onChange={[Function]}
        placeholder="(Min)"
        type="tel"
      />
      <button
        aria-label="timer-start"
        onClick={[Function]}
      >
        Start
      </button>
    </span>
    <span>
      More than halfway there!
    </span>
    <div>
       
      <span
        aria-label="timer-display"
      >
        00:00
      </span>
      <span
        aria-label="pause-button"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={40}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </div>
    <span>
      <button
        aria-label="single-speed-button"
        onClick={[Function]}
      >
        1x
      </button>
      <button
        aria-label="one-and-half-speed"
        onClick={[Function]}
      >
        1.5
      </button>
      <button
        aria-label="double-speed-button"
        onClick={[Function]}
      >
        2x
      </button>
    </span>
  </div>
</div>
`;
